pkg/generate/code/resource_reference.go

```diff
@@ -179,6 +185,187 @@ func ReferenceFieldsPresent(
 	return iteratorsOut + returnOut
 }
 
+// ResolveReferencesForField produces Go code for accessing all references that
+// are related to the given concrete field, determining whether its in a valid
+// condition and updating the concrete field with the referenced value.
+// Sample code:
+//
+// ```
+// refVal := ""
+//
+//	if ko.Spec.TargetRef != nil && ko.Spec.TargetRef.From != nil {
+//		arr := ko.Spec.TargetRef.From
+//		if arr == nil || arr.Name == nil || *arr.Name == "" {
+//			return fmt.Errorf("provided resource reference is nil or empty")
+//		}
+//		namespacedName := types.NamespacedName{
+//			Namespace: namespace,
+//			Name:      *arr.Name,
+//		}
+//		obj := svcapitypes.Integration{}
+//		err := apiReader.Get(ctx, namespacedName, &obj)
+//		if err != nil {
+//			return err
+//		}
+//		var refResourceSynced, refResourceTerminal bool
+//		for _, cond := range obj.Status.Conditions {
+//			if cond.Type == ackv1alpha1.ConditionTypeResourceSynced &&
+//				cond.Status == corev1.ConditionTrue {
+//				refResourceSynced = true
+//			}
+//			if cond.Type == ackv1alpha1.ConditionTypeTerminal &&
+//				cond.Status == corev1.ConditionTrue {
+//				refResourceTerminal = true
+//			}
+//		}
+//		if refResourceTerminal {
+//			return ackerr.ResourceReferenceTerminalFor(
+//				"Integration",
+//				namespace, *arr.Name)
+//		}
+//		if !refResourceSynced {
+//			return ackerr.ResourceReferenceNotSyncedFor(
+//				"Integration",
+//				namespace, *arr.Name)
+//		}
```

Comment:
Seems to me that we could create a function in the `references.go.tpl` template file that handled this in a standard fashion:

```go
{{- objType := ( printf "%sapitypes.%s" .ReferencedServiceName .FieldConfig.References.Resource ) -}}
{{- if eq .FieldConfig.References.ServiceName "" -}}
{{- objType := ( printf "svcapitypes.%s" {{ .FieldConfig.References.Resource) -}}
{{ end -}}
// getReferencedResourceState_{{ .FieldConfig.References.Resource }} looks up whether a referenced resource
// exists and is in a ACK.ResourceSynced=True state. If the referenced resource does exist and is
// in a Synced state, returns nil, otherwise returns `ackerr.ResourceReferenceTerminalFor` or
// `ResourceReferenceNotSyncedFor` depending on if the resource is in a Terminal state.
func getReferencedResourceState_{{ .FieldConfig.References.Resource }}(
	obj *{{ objType }},
	name string, // the Kubernetes name of the referenced resource
) error {
	namespacedName := types.NamespacedName{
		Namespace: namespace,
		Name: name,
	}
	err := apiReader.Get(ctx, namespacedName, &obj)
	if err != nil {
		return err
	}
	var refResourceSynced, refResourceTerminal bool
	for _, cond := range obj.Status.Conditions {
		if cond.Type == ackv1alpha1.ConditionTypeResourceSynced &&
			cond.Status == corev1.ConditionTrue {
			refResourceSynced = true
		}
		if cond.Type == ackv1alpha1.ConditionTypeTerminal &&
			cond.Status == corev1.ConditionTrue {
			return ackerr.ResourceReferenceTerminalFor(
				"{{ .FieldConfig.References.Resource }}",
				namespace, name)
		}
	}
	if !refResourceSynced {
		return ackerr.ResourceReferenceNotSyncedFor(
			"{{ .FieldConfig.References.Resource }}",
			namespace, name)
	}
	return nil
}
```

We'd generate the above functions for each referenced field. So, imagine we had 2 reference fields called `SubnetID` and `VPCID`, we'd generate two functions: `getReferencedResourceState_SubnetID()` and `getReferencedResourceState_VPCID()`.

You would then be able to call those functions from your Go code-generating functions instead of cluttering the Go code-generating functions with heavy indentation. It would also make things easier to debug and troubleshoot since the very common "look this up and see if the resource is synced" logic is isolated into a single function, specialized for that reference field.